ethcore/src/engines/validator_set/contract.rs

```diff
@@ -22,50 +22,48 @@ use ethereum_types::{H256, Address};
 use parking_lot::RwLock;
 use bytes::Bytes;
 
-use futures::Future;
-use native_contracts::ValidatorReport as Provider;
-
 use client::EngineClient;
 use header::{Header, BlockNumber};
 use machine::{AuxiliaryData, Call, EthereumMachine};
 
 use super::{ValidatorSet, SimpleList, SystemCall};
 use super::safe_contract::ValidatorSafeContract;
 
+use_contract!(validator_report, "ValidatorReport", "res/contracts/validator_report.json");
+
 /// A validator contract with reporting.
 pub struct ValidatorContract {
+	contract_address: Address,
 	validators: ValidatorSafeContract,
-	provider: Provider,
+	provider: validator_report::ValidatorReport,
 	client: RwLock<Option<Weak<EngineClient>>>, // TODO [keorn]: remove
```

Comment:
Actually `Option<Weak<>>` doesn't make any sense, since you can always do `Weak::new()` (will just return `None`) when you try to upgrade it.

Reply:
good catch!

Reply:
actually, I can'y do that, cause `EngineClient` is a trait.

```
cargo check -p ethcore
   Compiling ethcore v1.9.0 (file:///Users/marek/projects/ethcore/parity/ethcore)
error[E0277]: the trait bound `client::traits::EngineClient: std::marker::Sized` is not satisfied
  --> ethcore/src/engines/validator_set/contract.rs:48:24
   |
48 |    client: RwLock::new(Weak::new()),
   |                        ^^^^^^^^^ `client::traits::EngineClient` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `client::traits::EngineClient`
   = note: required by `<std::sync::Weak<T>>::new`
```